reflectionserializer/manager.go

```diff
@@ -0,0 +1,795 @@
+	case reflect.Uint8:
+		tmp, err := buffer.ReadUint8()
+		if err != nil {
+			return nil, err
+		}
+		return tmp, nil
+	case reflect.Uint16:
+		tmp, err := buffer.ReadUint16()
+		if err != nil {
+			return nil, err
+		}
+		return tmp, nil
+	case reflect.Uint32:
+		tmp, err := buffer.ReadUint32()
+		if err != nil {
+			return nil, err
+		}
+		return tmp, nil
+	case reflect.Uint64:
+		tmp, err := buffer.ReadUint64()
+		if err != nil {
+			return nil, err
+		}
+		return tmp, nil
+	case reflect.Uint:
+		tmp, err := buffer.ReadUint64()
+		if err != nil {
+			return nil, err
+		}
+		return uint(tmp), nil
+	case reflect.Float32:
+		tmp, err := buffer.ReadFloat32()
+		if err != nil {
+			return nil, err
+		}
+		return tmp, nil
+	case reflect.Float64:
+		tmp, err := buffer.ReadFloat64()
+		if err != nil {
+			return nil, err
+		}
+		return tmp, nil
+	case reflect.String:
+		tmp, err := ReadLen(fieldMetadata.LengthPrefixType, buffer)
+		if err != nil {
+			return nil, err
+		}
+		bytesRead, err := buffer.ReadBytes(tmp)
+		if err != nil {
+			return nil, err
+		}
+		restoredString := string(bytesRead)
+		return restoredString, nil
+	case reflect.Array:
+		return m.deserializeArray(valueType, fieldMetadata, buffer)
+	case reflect.Slice:
+		return m.deserializeSlice(valueType, fieldMetadata, buffer)
+	case reflect.Map:
+		return nil, fmt.Errorf("native map type is not supported. use orderedmap instead")
+	case reflect.Ptr:
+		return m.deserializePointer(valueType, fieldMetadata, buffer)
+	case reflect.Struct:
+		return m.deserializeStructure(valueType, fieldMetadata, buffer)
+	case reflect.Interface:
+		return m.deserializeInterface(valueType, fieldMetadata, buffer)
+	}
+	return nil, nil
+}
+
+func (m *SerializationManager) deserializeInterface(valueType reflect.Type, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) (interface{}, error) {
+	// if interface value can be nil, write first byte to indicate whether it has value
+	if fieldMetadata.AllowNil {
+		nilPointer, err := buffer.ReadByte()
+		if err != nil {
+			return nil, err
+		}
+		// if interface prefix is 0 then set value of the interface to nil
+		if nilPointer == 0 {
+			return nil, nil
+		}
+	}
+	encodedType, err := buffer.ReadUint32()
+	if err != nil {
+		return nil, err
+	}
+	implementationType, err := m.DecodeType(encodedType)
+	if err != nil {
+		return nil, err
+	}
+	if !implementationType.Implements(valueType) {
+		return nil, fmt.Errorf("couldn't deserialize interface: %s must implement interface %s", implementationType, valueType)
+	}
+	return m.DeserializeType(implementationType, fieldMetadata, buffer)
+}
+
+func (m *SerializationManager) deserializeStructure(valueType reflect.Type, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) (interface{}, error) {
+	// handle time struct individually
+	if valueType == reflect.TypeOf(time.Time{}) {
+		restoredTime, err := buffer.ReadTime()
+		if err != nil {
+			return nil, err
+		}
+		return restoredTime, nil
+	}
+
```

Comment:
Not really, because there we check `valueType.Kind()` of an object and here, knowing that it's a struct, we check if it's an instance of `time.Time` structure.